persistence-api/src/main/java/io/stargate/db/datastore/PersistenceBackedDataStore.java

```diff
@@ -1,81 +1,178 @@
+        .prepare(query.queryStringForPreparation(), parameters())
+        .thenApply(prepared -> query.withPreparedId(prepared.statementId));
   }
 
   @Override
-  public CompletableFuture<ResultSet> batch(List<String> queries) {
+  public CompletableFuture<ResultSet> execute(
+      BoundQuery query, UnaryOperator<Parameters> parametersModifier) {
     long queryStartNanos = System.nanoTime();
-    List<Statement> persistenceStatements =
-        queries.stream().map(SimpleStatement::new).collect(Collectors.toList());
-    return batch(persistenceStatements, BatchType.LOGGED, parameters, queryStartNanos);
+    Parameters executeParameters = parametersModifier.apply(parameters());
+    validateExecuteParameters(executeParameters);
+
+    CompletableFuture<ResultSet> future = new CompletableFuture<>();
+    if (query.bounded().query().preparedId().isPresent()) {
+      executeWithRetry(query, executeParameters, queryStartNanos, future);
+    } else if (options.alwaysPrepareQueries()) {
+      prepareAndRetry(query.bounded(), executeParameters, queryStartNanos, future);
+    } else {
+      doExecute(query, executeParameters, queryStartNanos, future, future::completeExceptionally);
+    }
+    return future;
+  }
+
+  private void executeWithRetry(
+      BoundQuery query,
+      Parameters executeParameters,
+      long queryStartNanos,
+      CompletableFuture<ResultSet> future) {
+    doExecute(
+        query,
+        executeParameters,
+        queryStartNanos,
+        future,
+        ex -> {
+          if (ex instanceof PreparedQueryNotFoundException) {
+            // This could happen due to a schema change between the statement preparation and now,
+            // as some schema change can invalidate preparation.
+            logger.debug(
+                "Prepared statement (id={}) was invalid when executed. This can happen due to a "
+                    + "conflicting schema change. Will re-prepare and retry.",
+                ((PreparedQueryNotFoundException) ex).id);
+            prepareAndRetry(query.bounded(), executeParameters, queryStartNanos, future);
+          } else {
+            future.completeExceptionally(ex);
+          }
+        });
+  }
+
+  private void doExecute(
+      BoundQuery query,
+      Parameters executeParameters,
+      long queryStartNanos,
+      CompletableFuture<ResultSet> successFuture,
+      Consumer<Throwable> onException) {
+    Statement statement = toPersistenceStatement(query);
+    connection
+        .execute(statement, executeParameters, queryStartNanos)
+        .thenAccept(
+            r ->
+                successFuture.complete(
+                    PersistenceBackedResultSet.create(connection, r, statement, executeParameters)))
+        .exceptionally(
+            ex -> {
+              onException.accept(ex);
+              return null;
+            });
+  }
+
+  private void prepareAndRetry(
```

Comment:
Actually it looks like it's going to retry until either a) it succeeds or b) throws an exception other than `PreparedQueryNotFoundException`.

`prepareAndRetry` calls `executeWithRetry` which tries to execute and if it catches `PreparedQueryNotFoundException` will call `prepareAndRetry`, so on and so forth.

Reply:
One would hope we never need retry more than once, but there is nothing strongly preventing it.

The reason this retry exists in the first place is that in rare case (on some schema changes), the server can invalidate a prepared statement. Typically, if you prepared the statement `SELECT * FROM t ...`, then if a new column is added to table t, that statement will be invalidated because the result set changes (it has one more column).
Anyway, in general, if this ever happens, you re-prepare and you should be good. But in theory, nothing prevents another schema change to interfere and requiring a new prepare.

So yet, that code retries until either our prepare sticks, or some other problem occurs, and that's the intended behavior of said code.

I'll remark btw that I'm not adding a new behavior with that patch here (if that was the worry) That patch only move the existing logic (which previously was in the removed `PersistenceBackedPreparedStatement` file) and adapt it slightly to changes of this patch, which it why it looks new-to-this-patch.